### PR TITLE
feat(NgComponentOutlet): add NgComponentOutlet directive

### LINKED ISSUE #9599: Proposal: Declarative Dynamic Components
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

As discussed in #8817, there is currently no convenient way to render a dynamic component. ```viewContainerRef.createComponent``` is too low-level for most use cases and requires to implement the update and destroy hook manually. This likely leads to bugs as well as suboptimal performance.

**Expected/desired behavior**

I suspect that ```createComponent``` will mostly be used to render a dynamic component at a certain position of a template. If this assumption holds true, then the same behaviour could be achieved with html code like the following:

```html
<template [ngComponent]="componentClass"
	[prop1]="propValue1"
	[ngProperties]="propertyValues">
```

When the code above is rendered, it would first resolve the component for the componentClass and instantiate the component. Then it would set all the regular properties. Since some properties may not be known at compile time, it would also be possible to provide an object with additional properties. At the end, the component is added directly to the DOM (without a wrapper element).

This could then also be used to render a list of dynamic components like this:

```html
<template
	*ngFor="let c of components"
	[ngComponent]="c.componentClass"
	[prop1]="c.propValue1"
	[ngProperties]="c.propertyValues">
```

Whenever the component class changes, the current component would replaced with a component instance of the new type.

I have chosen ```<template>``` for the placeholder because the placeholder itself must not be added to the DOM. I am not sure if this is the ideal element for the placeholder though.

This proposal could also be used as a replacement for ngInclude (#7596 and #2753).

### LINKED ISSUE #11168: [feature] NgComponentOutlet
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report => search github for a similar issue or PR before submitting
[ x ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
None

**Expected/desired behavior**
A directive that is able to create components dynamically, the same way as `NgTemplateOutlet` created embedded view dynamically.
#### Pros
- Remove the boilerplate in a component. Developer does not need to query (`ViewChild`) and manually create the component. (`ComponentFactoryResolver`, `ViewContainerRef`)
- Expressed in the template.
- Simple
- Aligned with NgTemplateOutlet, same idea same concept
#### Cons
- Host attribute bindings are not expressed in a template. Not much of a con, as its expected and `ComponentRef` can be used to set values on the instance...
- Getting a reference to the returned `ComponentRef` is via callback only.
- more ?

**Directive**

``` ts
@Directive({
  selector: '[swapCmp]'
})
export class SwapComponentDirective {
  private component: any;

  constructor(private cfr: ComponentFactoryResolver,
              private vcRef: ViewContainerRef,
              private tRef: TemplateRef<Object>) {
  }

  @Input() swapCmpBindings: ResolvedReflectiveProvider[];
  @Input() swapCmpInjector: Injector;
  @Input() swapCmpProjectables: any[][];

  @Output() onCreate: EventEmitter<ComponentRef<any>> = new EventEmitter<ComponentRef<any>>(false);

  @Input() set swapCmp(component: any) {
    this.component = component;
    this.vcRef.clear();
    if (this.component) {
      let injector = this.swapCmpInjector || this.vcRef.parentInjector;

      if (Array.isArray(this.swapCmpBindings) && this.swapCmpBindings.length > 0) {
        injector = ReflectiveInjector.fromResolvedProviders(this.swapCmpBindings, injector);
      }

      const cmpRef = this.vcRef.createComponent(
        this.cfr.resolveComponentFactory(component),
        this.vcRef.length,
        injector,
        this.swapCmpProjectables
      );

      cmpRef.changeDetectorRef.detectChanges();

      this.onCreate.emit(cmpRef);
    }
  }
}
```

```
 <template [swapCmp]="ctx.component" [swapCmpBindings]="ctx.bindings"   [swapCmpInjector]="ctx.injector" [swapCmpProjectables]="ctx.projectableNodes"></template>
```

> The name here is SwapComponent but it just as an example, took it from my [code base](https://github.com/shlomiassaf/angular2-modal/blob/master/src/components/angular2-modal/components/swap-component.directive.ts)
- A Developer provides a component, changing the component will re-create.
- Injector is optional, if not supplied the current injector is used.
- Bindings are used to allow additional providers to the injector (something like `viewProviders`)
- In case `ng-content` is needed the developer can supply projectableNodes.

What do you think?

### PR DESCRIPTION
**Pull request by shlomiassaf**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
See #11168 

**What is the new behavior?**
See #11168 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

This commits adds a new NgComponentOutlet directive that can be
used to dynamically create host views from a supplied component type/class.

Closes #11168
Closes #9599

### CODE REVIEW

File: modules/@angular/router/src/router.ts [hunk 1]
```diff
@@ -546,7 +546,7 @@ export class Router {
                   resolvePromise(false);
                 }
               },
-            (e: any) => {
+              (e: any) => {
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: I think with commit is unrequired if you rebase on master

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ */
+@Directive({selector: 'template[ngComponentOutlet]'})
+export class NgComponentOutlet implements OnChanges {
+  @Input() ngComponentOutlet: Type<any>;
+  @Input() ngOutletInjector: Injector;
+  @Input() ngOutletProviders: Provider[];
+  @Input() ngOutletProjectableNodes: any[][];
+
+  @Output()
+  ngOutletCreated: EventEmitter<ComponentRef<any>> = new EventEmitter<ComponentRef<any>>(false);
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: `: EventEmitter<ComponentRef<any>>` is not required here. I would personally drop it. Up to you.

Comment (shlomiassaf) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: Use type inference you mean?
Old habits... I will remove it once I get home

Comment (vicb) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: I also usually favor explicit but in some cases I find it hurt readability

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * <template [ngComponentOutlet]="componentTypeExpression"
+ *           [ngOutletInjector]="injectorExpression"
+ *           [ngOutletProviders]="providersExpression"
+ *           [ngOutletProjectableNodes]="projectableNodesExpression">
+ * </template>
+ * ```
+ *
+ * @experimental
+ */
+@Directive({selector: 'template[ngComponentOutlet]'})
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: remove `template` and document the short syntax ?

`<ng-container *ngComponentOutlet="..."></ng-container>`

Comment (shlomiassaf) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: There no gain in using the short syntax since the element will never be used.
The only thing that get's rendered is the component.

I chose to enforce `template`so this confusion can be avoided.

I had a discussion about this with @alexeagle , he also supports that.
I believe this should be in `NgTemplateOutlet` as well.

However, if i'm missing something we can change this.

File: modules/@angular/router/src/router.ts (hunk 1 above)
Comment (shlomiassaf) [inline modules/@angular/router/src/router.ts]: Yep, someone got me to it :)

File: modules/@angular/common/src/directives/ng_component_outlet.ts (hunk 3 above)
Comment (vicb) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: > he also supports that

but IE doesn't :)

**Comment by shlomiassaf** [comment]
@vicb Should I do something regarding the 'pr_action: cleanup'?

The [TRIAGE_AND_LABELS.md](https://github.com/angular/angular/blob/master/TRIAGE_AND_LABELS.md) says:
```
pr_action: cleanup -- more work is needed from the current assignee.
```

Who's the 'assignee'?

**Comment by vicb** [comment]
- we won't merge new features into master for for next few weeks only bug fixes,
- you still have not address my comment for IE not supporting `template`

**Comment by shlomiassaf** [comment]
@vicb Good point on IE, will fix.

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+export class NgComponentOutlet implements OnChanges {
+  @Input() ngComponentOutlet: Type<any>;
+  @Input() ngOutletInjector: Injector;
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: all input must start with the `ngComponentOutlet` prefix

**Comment by vicb** [comment]
Could you please rebase ?

**Review by vicb — CHANGES_REQUESTED** [review]

**Comment by shlomiassaf** [comment]
@vicb almost done, one issue though...
I couldn't pass a test where using the sugared syntax `*` I can use an `@Output` on the structural directive.

There is the `ngComponentOutletCreated` `@Output` which I couldn't get working on the tests, I couldn't find any reference in the angular code either....

This:
```
`<div *ngComponentOutlet="currentComponent; providers:providers; created: onOutletCreate($event)"></div>`;
```

Will error this:
```
        Can't bind to 'ngComponentOutletCreated' since it isn't a known property of 'div'. ("<div [ERROR ->]*ngComponentOutlet="currentComponent; providers:providers; created: onOutletCreate($event)"

```

This:
`<div *ngComponentOutlet="currentComponent; providers:providers"
(ngComponentOutletCreated)="onOutletCreate($event)"></div>`;

Won't error but won't trigger...

If this is a limitation I want to note that `*` syntax also does not allow `exportAs` so it wont be possible to get the directive as a local template var and subscribe.
You could, I assume, use `@Self` or `@Host` to get the directive instance and subscribe.

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+      if (this.ngComponentOutlet) {
+        let injector = this.ngOutletInjector || this._viewContainerRef.parentInjector;
+
+        if (Array.isArray(this.ngOutletProviders) && this.ngOutletProviders.length > 0) {
+          injector = ReflectiveInjector.resolveAndCreate(this.ngOutletProviders, injector);
+        }
+
+        const cmpRef = this._viewContainerRef.createComponent(
+            this._cmpFactoryResolver.resolveComponentFactory(this.ngComponentOutlet),
```
Comment (zoechi) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: I'm wondering why this doesn't cause issues when't it's called before `ngAfterViewInit()`?
Isn't `ngOnChanges()` called before `ngAfterViewInit()`, at least the first time?
You would obviously have run into issues, so I guess it's safe to assume it's working fine.
Any hint what I might misunderstand?

Comment (shlomiassaf) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: What issues are you referring to?

`this.ngComponentOutlet` is checked in line 78, other null issues you see that i'm missing?

Comment (zoechi) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: I was sure that `viewContainerRef.createComponent()` can't be used before `ngAfterViewInit()`and was wondering why it seems to work for you.

File: modules/@angular/common/src/directives/ng_component_outlet.ts (hunk 4 above)
Comment (shlomiassaf) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: @vicb any thoughts?

File: modules/@angular/common/src/directives/ng_component_outlet.ts (hunk 5 above)
Comment (shlomiassaf) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: @zoechi 

This test passes: 
```ts
    it('should work when a component was set before ngOnInit', async(() => {
         let fixture = TestBed.createComponent(TestComponent);

         fixture.componentInstance.currentComponent = InjectedComponent;

         fixture.detectChanges();
         expect(fixture.nativeElement).toHaveText('foo');
       }));
```

For sure, `ngOnChanges` is called here **before** `ngOnInit` (thus `ngAfterViewInit`) and it triggers `viewContainerRef.createComponent()` since the component exists...

I was not aware of the limitation you mentioned, I experimented on [this plunker](http://plnkr.co/edit/2RzOXgCqo8HBoYgA3w4h?p=preview) it looks like this limitation does not exists at this point.

Am i'm missing something?

Comment (zoechi) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: @shlomiassaf thanks. Seems I'll have to investigate and update my mental model ...

**Comment by shlomiassaf** [comment]
@vicb @vsavkin I need some help with the unsupported feature in angular.
Sugared directives does no support `@Output`
Should I just ignore that?

**Comment by mhevery** [comment]
Ping @vicb / @vsavkin, can you help with reviewing this PR?

**Comment by shlomiassaf** [comment]
@mhevery / @vicb / @vsavkin,

Gentle reminder on this topic, we are currently block by #12121.

**TL;DR:** 
  * To support IE we need to support sugar syntax (no template tag)
  * Sugar syntax does not support @Output registration (the blocker...)
 
Can you advise how to move forward?

Should I ignore the lack of support for `@Output` when using sugar syntax?

File: modules/@angular/common/src/directives/ng_component_outlet.ts (hunk 3 above)
Comment (mhevery) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: actually, giving it a bit more thought this is even better:

```
<ng-container *ngOutlet="component: componentTypeExpression; injector: injExp; provider: proExp; projectable: proExp"></ng-container>

```

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+ *           [ngOutletProviders]="providersExpression"
+ *           [ngOutletProjectableNodes]="projectableNodesExpression">
+ * </template>
+ * ```
+ *
+ * @experimental
+ */
+@Directive({selector: 'template[ngComponentOutlet]'})
+export class NgComponentOutlet implements OnChanges {
+  @Input() ngComponentOutlet: Type<any>;
```
Comment (mhevery) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: rename to `ngOutletComponent`

File: modules/@angular/common/src/directives/ng_component_outlet.ts (hunk 4 above)
Comment (mhevery) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: If you do the rename above, then this name is fine.

File: modules/@angular/common/src/directives/ng_component_outlet.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+ *           [ngOutletInjector]="injectorExpression"
+ *           [ngOutletProviders]="providersExpression"
+ *           [ngOutletProjectableNodes]="projectableNodesExpression">
+ * </template>
+ * ```
+ *
+ * @experimental
+ */
+@Directive({selector: 'template[ngComponentOutlet]'})
+export class NgComponentOutlet implements OnChanges {
```
Comment (mhevery) [inline modules/@angular/common/src/directives/ng_component_outlet.ts]: should be renamed to `NgOutlet` since `Component` is a parameter of the outlet and should not be part of it.

**Review by mhevery — COMMENTED** [review]
I think this is a good PR and should be merged after we get the names right.

**Comment by shlomiassaf** [comment]
@mhevery 👍 

I already have a fixed version for sugar syntax but I do have blockers since `@Output` is not supported.

I will rename the component and post the PR.
Using sugar syntax will be fine but the output wont fire... until #12121 is fixed

**Comment by Luchillo** [comment]
What has become of this feature?

**Comment by shlomiassaf** [comment]
Sorry guys, long business trip, will get it done hopefully this weekend.

**Comment by dehru** [comment]
Thank you for doing this

**Comment by mhevery** [comment]
I have taken over this PR at #13383
